fe/fe-core/src/main/java/com/starrocks/service/InformationSchemaDataSource.java

```diff
@@ -256,63 +259,90 @@ public static TGetPartitionsMetaResponse generatePartitionsMetaResponse(TGetPart
+            public Element(String dbName, long dbId, Table table) {
+                this.dbName = dbName;
+                this.dbId = dbId;
+                this.table = table;
+            }
+            public long getTableId() {
+                return table.getId();
+            }
+            public String dbName;
+            public long dbId;
+            public Table table;
+        };
+        TreeSet<Element> sortedElements = new TreeSet<>(Comparator.comparing(Element::getTableId));
         for (String dbName : result.authorizedDbs) {
             Database db = GlobalStateMgr.getCurrentState().getLocalMetastore().getDb(dbName);
             if (db == null) {
                 continue;
             }
-            List<Table> allTables = GlobalStateMgr.getCurrentState().getLocalMetastore().getTables(db.getId());
-            for (Table table : allTables) {
-                try {
-                    Authorizer.checkAnyActionOnTableLikeObject(result.currentUser,
-                            null, dbName, table);
-                } catch (AccessDeniedException e) {
-                    LOG.warn("failed to check db: {} table: {} authorization", dbName, table, e);
-                    continue;
-                }
-                if (!table.isNativeTableOrMaterializedView()) {
-                    continue;
-                }
-                // only olap table/mv or cloud table/mv will reach here;
-                // use the same lock level with `SHOW PARTITIONS FROM XXX` to ensure other modification to
-                // partition does not trigger crash
-                Locker locker = new Locker();
-                locker.lockDatabase(db.getId(), LockType.READ);
-                try {
-                    OlapTable olapTable = (OlapTable) table;
-                    PartitionInfo tblPartitionInfo = olapTable.getPartitionInfo();
-                    // normal partition
-                    for (Partition partition : olapTable.getPartitions()) {
-                        for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
-                            TPartitionMetaInfo partitionMetaInfo = new TPartitionMetaInfo();
-                            partitionMetaInfo.setDb_name(dbName);
-                            partitionMetaInfo.setTable_name(olapTable.getName());
-                            genPartitionMetaInfo(db, olapTable, tblPartitionInfo, partition, physicalPartition,
-                                    partitionMetaInfo, false /* isTemp */);
-                            pList.add(partitionMetaInfo);
-                        }
+            List<Table> tables = GlobalStateMgr.getCurrentState().getLocalMetastore().getTables(db.getId());
+            for (Table table : tables) {
+                sortedElements.add(new Element(dbName, db.getId(), table));
+            }
+        }
+
+        long startTableIdOffset = request.isSetStart_table_id_offset() ? request.getStart_table_id_offset() : 0;
```

Comment:
shall move this condition before TreeSet, so in line 283, it is easy to determine if the table needs to be added into TreeSet or not.